source/core/index.ts

```diff
@@ -0,0 +1,1472 @@
+		});
+	}
+}
+
+export class CacheError extends RequestError {
+	constructor(error: Error, options: NormalizedOptions) {
+		super(error.message, error, options);
+		this.name = 'CacheError';
+	}
+}
+
+export class UploadError extends RequestError {
+	constructor(error: Error, options: NormalizedOptions, request: Request) {
+		super(error.message, error, options, request);
+		this.name = 'UploadError';
+	}
+}
+
+export class TimeoutError extends RequestError {
+	readonly timings: Timings;
+	readonly event: string;
+
+	constructor(error: TimedOutTimeoutError, timings: Timings, options: NormalizedOptions) {
+		super(error.message, error, options);
+		this.name = 'TimeoutError';
+		this.event = error.event;
+		this.timings = timings;
+	}
+}
+
+export class ReadError extends RequestError {
+	constructor(error: Error, options: NormalizedOptions, response: Response) {
+		super(error.message, error, options, response);
+		this.name = 'ReadError';
+	}
+}
+
+export class UnsupportedProtocolError extends RequestError {
+	constructor(options: NormalizedOptions) {
+		super(`Unsupported protocol "${options.url.protocol}"`, {}, options);
+		this.name = 'UnsupportedProtocolError';
+	}
+}
+
+const proxiedRequestEvents = [
+	'socket',
+	'abort',
+	'connect',
+	'continue',
+	'information',
+	'upgrade',
+	'timeout'
+];
+
+export default class Request extends Duplex implements RequestEvents<Request> {
+	['constructor']: typeof Request;
+
+	declare [kUnproxyEvents]: () => void;
+	declare _cannotHaveBody: boolean;
+	[kDownloadedSize]: number;
+	[kUploadedSize]: number;
+	[kBodySize]?: number;
+	[kServerResponsesPiped]: Set<ServerResponse>;
+	[kIsFromCache]?: boolean;
+	[kStartedReading]?: boolean;
+	[kCancelTimeouts]?: () => void;
+	[kResponseSize]?: number;
+	[kResponse]?: IncomingMessage;
+	[kRequest]?: ClientRequest;
+	_noPipe?: boolean;
+	_progressCallbacks: Array<() => void>;
+
+	declare options: NormalizedOptions;
+	declare requestUrl: string;
+	finalized: boolean;
+	redirects: string[];
+
+	constructor(url: string | URL, options: Options = {}, defaults?: Defaults) {
+		super({
+			// It needs to be zero because we're just proxying the data to another stream
+			highWaterMark: 0
+		});
+
+		this[kDownloadedSize] = 0;
+		this[kUploadedSize] = 0;
+		this.finalized = false;
+		this[kServerResponsesPiped] = new Set<ServerResponse>();
+		this.redirects = [];
+
+		// TODO: Remove this when targeting Node.js >= 12
+		this._progressCallbacks = [];
+
+		const unlockWrite = (): void => this._unlockWrite();
+		const lockWrite = (): void => this._lockWrite();
+
+		this.on('pipe', (source: Writable) => {
+			source.prependListener('data', unlockWrite);
+			source.on('data', lockWrite);
+
+			source.prependListener('end', unlockWrite);
+			source.on('end', lockWrite);
+		});
+
+		this.on('unpipe', (source: Writable) => {
+			source.off('data', unlockWrite);
+			source.off('data', lockWrite);
+
+			source.off('end', unlockWrite);
+			source.off('end', lockWrite);
+		});
```

Comment:
This is actually needed when people pass custom `options.body` (stream), so

```
got.post('https://example.com', {body: readableStream}).write('asdf');
```

throws (the current Got implementation allows this behavior).